### PR TITLE
【Hackathon No.24】add RFC for ChannelShuffle

### PR DESCRIPTION
**Pull request by BrilliantYuKaimin**
增加paddle.nn.ChannelShuffle设计文档。

### CODE REVIEW

**Comment by dingjiaweiww** [comment]
你的 PR 提交成功，感谢你对于开源项目的贡献，请检查 PR 提交格式和内容是否完备，具体请参考[示例](https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/20200301_api_design_for_quantile.md)和[模版](https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/api_design_template.md)。

**Comment by dingjiaweiww** [comment]
PR 格式检查通过，你的PR 将接受Paddle 专家以及开源社区的review，请及时关注PR 动态

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+```python
+def shuffle_unit(self, x, groups):
+    with tf.variable_scope('shuffle_unit'):
+        n, h, w, c = x.get_shape().as_list()
+        x = tf.reshape(x, shape=tf.convert_to_tensor([tf.shape(x)[0], h, w, groups, c // groups]))
+        x = tf.transpose(x, tf.convert_to_tensor([0, 1, 2, 4, 3]))
+        x = tf.reshape(x, shape=tf.convert_to_tensor([tf.shape(x)[0], h, w, c]))
+```
+
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_ChannelShuffle.md:74]: 请就这两种方案的实现方式的整体逻辑进行分步骤阐述

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+# 四、对比分析
+无论是C++实现还是组合API实现，其逻辑都是十分简单的，故考虑使用C++编写新的算子以期取得更高的效率。
+
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_ChannelShuffle.md:77]: 请就 torch/tensorflow 两种实现方案进行对比评价和对比分析，论述优势、劣势。

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+- **groups** (int) - 要把通道数分成的组数，必需整除通道数。
+- **data_format**(str) - 指定输入的数据格式，输出的数据格式将与输入保持一致，可以是"NCHW"和"NHWC"。N是批尺寸，C是通道数，H是特征高度，W是特征宽度。默认值："NCHW"。
+
+## 底层OP设计
+ChannelShuffle与飞桨中已有的PixelShuffle在操作上是类似的，参考PixelShuffle的OP设计来实现ChannelShuffle的底层OP。
+
+## API实现方案
+
+参考`paddle.nn.PixelShuffle`来实现`paddle.nn.ChannelShuffle`，顺便实现`paddle.nn.functional.channel_shuffle`。
+
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_ChannelShuffle.md:92]: 请对实现方案包含实现文件存放位置，代码内部逻辑进行步骤化的描述

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+# 六、测试和验收的考量
+
+考虑测试的情况：
+- 与PyTorch的结果的一致性；
+- 反向传播的正确性；
+- 错误检查：`groups`不合法或不整除通道数时能正确抛出异常。
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_ChannelShuffle.md:98]: 测试内容需要进一步参考这里：https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/dev_guides/api_contributing_guides/api_accpetance_criteria_cn.html#id1

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+在飞桨框架中增加ChannelShuffle 组网 API。
+
+## 3、意义
+
+飞桨将支持ChannelShuffle 组网 API。
+
+# 二、飞桨现状
+飞桨目前不支持此功能，但可以通过组合API的方式实现此功能。
+
```
[the remark this answers is not in the document]
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_ChannelShuffle.md]: 完成

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md (hunk 2 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_ChannelShuffle.md]: > 请就 torch/tensorflow 两种实现方案进行对比评价和对比分析，论述优势、劣势。

完成

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md (hunk 3 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_ChannelShuffle.md:143]: > 请对实现方案包含实现文件存放位置，代码内部逻辑进行步骤化的描述

完成

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md (hunk 4 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_ChannelShuffle.md]: > 测试内容需要进一步参考这里：https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/dev_guides/api_contributing_guides/api_accpetance_criteria_cn.html#id1

完成

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+我们以NCHW格式的4D张量为例来描述ChannelShuffle的实现逻辑：
+1. 把形为[N, C, H, W]的张量重塑成[N, g, C/g, H, W]的形状；
+2. 对第1维度（N为第0维度）和第2维度进行转置，得到形为[N, C/g, g, H, W]的张量；
+3. 把形为[N, C/g, g, H, W]的张量重塑为[N, C, H, W]的形状。
+
+# 四、对比分析
+- PyTorch在C++层面为ChannelShuffle设计了底层算子，执行效率高，但不利于（不参与框架设计）开发者从源码层面了解该API的行为。
+- 使用组合API的方式来实现ChannelShuffle是简洁易懂的，开发者可以直接从Python源码的层面来了解该API的行为，但在执行效率上可能会逊色于原生算子实现方案。
+
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_ChannelShuffle.md:97]: 此处的对比分析应当对比pytorch 和 tf 两种方案的优劣，此处两种方法差异不大，则可以直接说明分Channel形式来进行和两种方案一致的的设计

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+        pass
+```
+
+# 六、测试和验收的考量
+
+- 是否同时支持静态图和动态图；
+- 是否同时支持CPU和GPU平台；
+- 测试不同张量类型下的表现；
+- 对全部入参进行参数有效性和边界值测试，确定每个入参都可以正确生效；
+- 前向计算的正确性（与组合API实现比较、与PyTorch比较）；
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_ChannelShuffle.md:180]: 可以在本地和pytorch实现比较，在test由于CI环境没有pytorch，需要使用 numpy 实现进行比较。

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md [hunk 8]
```diff
@@ -0,0 +1,197 @@
...
+class ChannelShuffle(Layer):
+    def __init__(self, groups, data_format="NCHW", name=None):
+        pass
+
+    def forward(self, x):
+        return functional.channel_shuffle(x, self._groups,
+                                          self._data_format, self._name)
+
+    def extra_repr(self):
+        pass
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_ChannelShuffle.md:171]: 这个方法是否需要？

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+# 六、测试和验收的考量
+
+- 是否同时支持静态图和动态图；
+- 是否同时支持CPU和GPU平台；
+- 测试不同张量类型下的表现；
+- 对全部入参进行参数有效性和边界值测试，确定每个入参都可以正确生效；
+- 前向计算的正确性（与组合API实现比较、与PyTorch比较）；
+- 反向计算的正确性；
+- 当传入的`group`不合法（不是正整数、不整除通道数）时会抛出异常并有友好的提示。
+
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_ChannelShuffle.md:183]: 错误检查：data_format不合法，x不是tensor，x维度检查等

**Review by shiyutang — CHANGES_REQUESTED** [review]

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md (hunk 8 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_ChannelShuffle.md:171]: 其他API均有此方法。

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md (hunk 6 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_ChannelShuffle.md]: 完成

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md (hunk 7 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_ChannelShuffle.md]: 完成

File: rfcs/APIs/20220317_api_design_for_ChannelShuffle.md (hunk 9 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_ChannelShuffle.md:187]: 完成

**Review by shiyutang — APPROVED** [review]
LGTM
